### ISSUE TITLE
Support for OCI registries in Flux plugin

### ISSUE THREAD

**Issue by gfichtenholt** [vmware-tanzu/kubeapps#5007]
This is a long-running issue I started a little over a month ago.
Flux CD added support for OCI registries recently
https://github.com/fluxcd/flux2/pull/2597

My work is to expose this functionality to kubeapps end user via flux plugin.

Areas that I worked on so far:
- Basic CRUD support for OCI registries, including caching of repos and charts in redis
  (like an index for regular helm repositories)
  U(pdate) implementation still pending.
- Introduce an API for repository lister (catalog). Implementation so far is
  docker api v2 compatible registry
  https://github.com/opencontainers/distribution-spec/blob/main/spec.md#api using
  [ORAS v2 libraries](https://github.com/oras-project/oras-go). Works for GitHub (ghcr.io), others TBD
  Full discussion here: https://github.com/vmware-tanzu/kubeapps/pull/4932. I stopped
  when I got blocked by https://github.com/oras-project/oras-go/issues/196 below. Once I verify it is fixed and will pick up
  where I left off

Areas which I have not started yet:
- Support other kinds of repositories, such as Harbor, Azure CR, AWS CR, etc.
- Support for k8s docker registry credentials (OCI registries may use them)
- Support for InsecureTLS flag
- unit tests. I wrote one integration test so far which is testing (back)end-2-end,
  because I wanted to see what flux engine really does on the server-side
  I know I need to do the client-side tests.

Currently outstanding issues that will require verification and/or code changes when fixed:
-  Add support for insecure registries in HelmRepositories of Type oci https://github.com/fluxcd/source-controller/issues/807
-  Flux fails to login to HTTP OCI registry where helm CLI logs in https://github.com/fluxcd/source-controller/issues/805
-  Support last parameter for listing catalog and tags  https://github.com/oras-project/oras-go/issues/196

I plan to be updating this issue as I make further progress with OCI support in flux plugin

**Comment by ppbaena** [comment]
Hi Greg, could you update the current status of this issue and next steps planned?

**Comment by gfichtenholt** [comment]
Current status: in progress.
Items left to do before I can mark this Closed:
- Update operation implementation for OCI repos + integration tests. 
  Time estimate: 2 days
- Production code cleanup (removing instrumentation via log.Info()). 
  Other tidying up: 2 days
- Integration tests: use cases involving TLS: 2 days
- Regular unit tests (i.e. when k8s isn't running). 
  I have so far only written and relied on integration tests running locally (not in CI) 
  which was fine until a point. Now I need to write at least some rudimentary CRUD tests that will be run on CI
  Time estimate: 2 days min. The more time I am able to spend the more tests I can write, including some 
  edge scenarios

These items are optional and can be addressed at any point in the future:
- Integration tests: remove runtime dependency on github ghcr.io as a test registry host, 
  in favor of a locally running registry. Blocked by https://github.com/fluxcd/source-controller/issues/805
- Add support for insecure registries in HelmRepositories. Blocked by https://github.com/fluxcd/source-controller/issues/807  
- Support for non-docker api v2 compatible OCI registries, such as Harbor, Azure CR, AWS CR, etc.

**Comment by gfichtenholt** [comment]
Current status:
- Update operation implementation for OCI repos + integration tests is blocked by 
https://github.com/fluxcd/source-controller/issues/839

**Comment by gfichtenholt** [comment]
just noticed flux intrduced a new source type in latest release https://fluxcd.io/docs/components/source/ocirepositories/ . I will need to spend some time to understand if/how this affects us.

**Comment by gfichtenholt** [comment]
This PR is still in progress with lots of changes coming in so adding this as comment here rather than creating a new issue: The current redis cache for helm OCI repos is partially broken. It loads the initial state into the cache, but when the state changes, such as:

- a new chart is added to repo
- an existing chart is removed from repo
- a new chart version (aka tag) is added to a chart 
- an existing tag is removed from a chart

Basically any change on remote server, the cache will get out of sync. The reason is that fluxcd chose not to implement a feature where the caller gets notified when source on remote changes (ref https://github.com/fluxcd/source-controller/issues/839). I am currently considering different fixes for this, such as:

- disable cache for helm OCI repositories
- implement logic to poll remote for changes and react (i.e. what flux is already doing)
- manual resync, including delete/create of helm OCI repo

will update this PR as I narrow the solution down

**Comment by gfichtenholt** [comment]
I am leaning towards disabling the repository cache for helm OCI repositories. This is the redis cache that contains mapping for repoName => list of models.Chart objects. Chart cache (i.e. each chart .tgz bytes) will still be enabled. 
The practical implication is that any call to `GetAvailablePackageSummaries` or `GetAvailablePackageVersions` will be implemented having to invoke the remote server for OCI repos. 
Why this choice? 
- There is a (remote) possibility that flux may enhance their support for OCI with the feature I asked for 
- For now, it is consistent with helm spec https://helm.sh/docs/topics/registries/#deprecated-features-and-strict-naming-policies. 
- The caching of the repo metadata is a nice to have for kubeapps, not a MUST, like the catalog of charts for repo. It just decreases latency of an API call
- I would rather have the feature disabled until a clean fix materializes than have it be broken. Having a clean fix is just a matter of time and priorities
I am currently working on having this in a PR. It will take me a couple of days. In the meantime, comments on this are welcome.

**Comment by gfichtenholt** [comment]
fyi, flux [v. 33](https://github.com/fluxcd/flux2/releases/tag/v0.33.0) adds [HelmRepository.spec.provider](https://fluxcd.io/docs/components/source/helmrepositories/#provider). I will need to find out more about it and whether it affects our code or not

**Comment by gfichtenholt** [comment]
Just FYI, this feature mentioned above appears to be something flux authors call "auto-login" or
"contextual login" that basically means using a "secured" HelmRepository without
a secret ref. It is important to understand that in order for this scenario to work,
the workload (i.e. flux controllers and kubeapps) needs to actually run on GKE cluster.
Another way to put it is "Workload Identity can't be used by Pods running on the host network"
ref https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#limitations

If I had to summarize this approach in one sentence it would something like "instead of
configuring secrets based off of as service access tokens or service JSON keys,
per helm repository, one configures flux source-controller pod with a certain
'workload identity' which is provided by flux to GKE when needed for authentication"

**Comment by castelblanque** [comment]
Thanks for the info, Greg, that makes it more understandable. The term "contextual login" or "contextual auth" looks like a good fit for referring to it.

Wondering how we can manage the feature from the UI point of view. For example: user creating a `HelmRepository` from Kubeapps UI could select the `provider` from e.g. a dropdown. Initially it could be a fixed list of the options in [the `provider` spec](https://fluxcd.io/docs/components/source/helmrepositories/#provider).
But for the experience to be optimal, the allowed options in the dropdown could be only two: 1) `generic` and 2) the provider in which Flux machinery runs only (e.g. `gcp`)?
Is the cloud provider something that we could find out from within Kubeapps?
In the multicluster scenario, the UI in which the repo would be created might not have access to the backend of the actual Flux running.

**Comment by gfichtenholt** [comment]
Haven't thought this far ahead yet.  As a first cut I was thinking of having users pick from a fixed list. I haven't done enough research to see if it's possible to find out automatically. Flux guys chose to put the burden on the user, so there is that

**Comment by gfichtenholt** [comment]
Status update: took 3 full days but I have now verified that the feature works as advertised in flux. That is Having the field spec.provider set to "gcp" is sufficient to make the HelmRepository work without a secret ref on GCP. Pretty neat.
That is, once you set up your GKE cluster to work as need, which is what took me the vast majority of the time.  Thinking about how to set up flux plugin integration test for this. Will probably be guarded by a flag. Unset by default. If set, rely on some existing GKE cluster that is pre-created and pre-configured by someone prior to running the test

**Comment by gfichtenholt** [comment]
I believe this issue can be now closed. All planned features have been implemented. If any new features rise we shall open separate PR targeting those
